File: 3rd Semester (C#)/Lab4/Banks.Test/BanksTests.cs
using Banks.Entities;
using Banks.Tools;
using Xunit;

namespace Banks.Test;

public class BanksTests
{
    private readonly CentralBank _centralBank = CentralBank.GetCentralBank();

    [Fact]
    public void BankCreation()
    {
        const string name = "Bank1";
        const double depositInterest = 10;
        const double creditComission = 100;
        const double debitComission = 5;
        const double doubtfulClientLimit = 50000;

        Bank new_bank = _centralBank.CreateNewBank(name, depositInterest, creditComission, debitComission, doubtfulClientLimit);

        Assert.Contains(new_bank, _centralBank.Banks);
        Assert.Equal(new_bank.Name, name);
        Assert.Equal(new_bank.DepositInterest, depositInterest);
        Assert.Equal(new_bank.CreditComission, creditComission);
        Assert.Equal(new_bank.DebitComission, debitComission);
        Assert.Equal(new_bank.DoubtfulClientLimit, doubtfulClientLimit);
    }

    [Fact]
    public void CreatingClientAndAddingToBank()
    {
        const string bank_name = "Bank2";
        const double depositInterest = 10;
        const double creditComission = 100;
        const double debitComission = 5;
        const double doubtfulClientLimit = 50000;

        Bank bank = _centralBank.CreateNewBank(bank_name, depositInterest, creditComission, debitComission, doubtfulClientLimit);

        const string client_name = "Ivan";
        const string client_surname = "Ivanov";

        ClientBuilder clientBuilder = new ();
        clientBuilder.SetFullName(new Models.FullName(client_name, client_surname));
        Client client = clientBuilder.CreateClient();

        bank.AddNewClient(client);

        Assert.Equal(client_name, client.FullName.Name);
        Assert.Equal(client_surname, client.FullName.Surname);
        Assert.Null(client.PassportData);
        Assert.Null(client.Address);
        Assert.Contains(client, bank.Clients);
    }

    [Fact]
    public void CreatingClientsAccounts()
    {
        const string bank_name = "Bank3";
        const double depositInterest = 10;
        const double creditComission = 100;
        const double debitComission = 5;
        const double doubtfulClientLimit = 50000;

        Bank bank = _centralBank.CreateNewBank(bank_name, depositInterest, creditComission, debitComission, doubtfulClientLimit);

        const string client_name = "Ivan";
        const string client_surname = "Ivanov";

        ClientBuilder clientBuilder = new ();
        clientBuilder.SetFullName(new Models.FullName(client_name, client_surname));
        Client client = clientBuilder.CreateClient();

        bank.AddNewClient(client);

        var deposit_period = TimeSpan.FromDays(60);
        const double money_amount = 100000;
        DebitAccount debitAccount = bank.CreateDebitAccount(client.Id);
        CreditAccount creditAccount = bank.CreateCreditAccount(client.Id);
        DepositAccount depositAccount = bank.CreateDepositAccount(client.Id, deposit_period, money_amount);

        Assert.Contains(debitAccount, client.Accounts);
        Assert.Contains(creditAccount, client.Accounts);
        Assert.Contains(depositAccount, client.Accounts);
    }

    [Fact]
    public void ReplenishFromAccount()
    {
        const string bank_name = "Bank4";
        const double depositInterest = 10;
        const double creditComission = 100;
        const double debitComission = 5;
        const double doubtfulClientLimit = 50000;

        Bank bank = _centralBank.CreateNewBank(bank_name, depositInterest, creditComission, debitComission, doubtfulClientLimit);

        const string client_name = "Ivan";
        const string client_surname = "Ivanov";

        ClientBuilder clientBuilder = new ();
        clientBuilder.SetFullName(new Models.FullName(client_name, client_surname));
        Client client = clientBuilder.CreateClient();

        bank.AddNewClient(client);

        const double money_amount = 100000;
        const double money_amount_twice = 200000;
        const double ZeroMoney = 0;
        DebitAccount debitAccount = bank.CreateDebitAccount(client.Id);

        Assert.Equal(debitAccount.Money, ZeroMoney);

        _centralBank.ReplenishAccount(debitAccount.Id, money_amount);
        Assert.Equal(debitAccount.Money, money_amount);

        _centralBank.ReplenishAccount(debitAccount.Id, money_amount);
        Assert.Equal(debitAccount.Money, money_amount_twice);
    }

    [Fact]
    public void WithdrawFromAccount()
    {
        const string bank_name = "Bank5";
        const double depositInterest = 10;
        const double creditComission = 100;
        const double debitComission = 5;
        const double doubtfulClientLimit = 50000;

        Bank bank = _centralBank.CreateNewBank(bank_name, depositInterest, creditComission, debitComission, doubtfulClientLimit);

        const string client_name = "Ivan";
        const string client_surname = "Ivanov";

        ClientBuilder clientBuilder = new ();
        clientBuilder.SetFullName(new Models.FullName(client_name, client_surname));
        Client client = clientBuilder.CreateClient();

        bank.AddNewClient(client);

        const double repl_money_amount = 100000;
        const double withdraw_money_amount = 40000;
        const double expected_money_left = repl_money_amount - withdraw_money_amount;

        DebitAccount debitAccount = bank.CreateDebitAccount(client.Id);
        _centralBank.ReplenishAccount(debitAccount.Id, repl_money_amount);

        _centralBank.WithdrawFromAccount(debitAccount.Id, withdraw_money_amount);
        Assert.Equal(expected_money_left, debitAccount.Money);
    }

    [Fact]
    public void TransferBetweenAccounts()
    {
        const string bank_name = "Bank6";
        const double depositInterest = 10;
        const double creditComission = 100;
        const double debitComission = 5;
        const double doubtfulClientLimit = 50000;

        Bank bank = _centralBank.CreateNewBank(bank_name, depositInterest, creditComission, debitComission, doubtfulClientLimit);

        const string client1_name = "Ivan";
        const string client1_surname = "Ivanov";

        ClientBuilder client1Builder = new ();
        client1Builder.SetFullName(new Models.FullName(client1_name, client1_surname));
        Client client1 = client1Builder.CreateClient();

        const string client2_name = "Name";
        const string client2_surname = "Surname";

        ClientBuilder client2Builder = new ();
        client2Builder.SetFullName(new Models.FullName(client2_name, client2_surname));
        Client client2 = client2Builder.CreateClient();

        bank.AddNewClient(client1);
        bank.AddNewClient(client2);

        const double repl_money_amount = 100000;
        const double transfer_money_amount = 30000;
        const double expected_money_left = repl_money_amount - transfer_money_amount;

        DebitAccount debitAccount1 = bank.CreateDebitAccount(client1.Id);
        DebitAccount debitAccount2 = bank.CreateDebitAccount(client2.Id);
        _centralBank.ReplenishAccount(debitAccount1.Id, repl_money_amount);

        _centralBank.TransferMoneyBetweenAccounts(debitAccount1.Id, debitAccount2.Id, transfer_money_amount);
        Assert.Equal(expected_money_left, debitAccount1.Money);
        Assert.Equal(transfer_money_amount, debitAccount2.Money);
    }
}

File: 4th Semester (Java)/lab-1/src/main/java/test/BanksTests.java
package test;

import accounts.CreditAccount;
import accounts.DebitAccount;
import accounts.DepositAccount;
import bank.Bank;
import bank.CentralBank;
import client.Client;
import client.ClientBuilder;
import models.FullName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Assertions;

import java.time.Duration;

/**
 * Class for Banks testing.
 */
public class BanksTests {
    /**
     * Central bank for management of the general banking system.
     */
    private final CentralBank centralBank = CentralBank.getCentralBank();

    /**
     * Testing if new bank is in central bank system, if bank has correct name, if all commissions, interests and limits are set correctly.
     */
    @Test
    public void bankCreation() {
        final String name = "Bank1";
        final double depositInterest = 10;
        final double creditCommission = 100;
        final double debitCommission = 5;
        final double doubtfulClientLimit = 50000;

        Bank newBank = centralBank.createNewBank(name, depositInterest, creditCommission, debitCommission, doubtfulClientLimit);

        Assertions.assertTrue(centralBank.getBanks().contains(newBank));
        Assertions.assertEquals(newBank.getName(), name);
        Assertions.assertEquals(newBank.getDepositInterest(), depositInterest);
        Assertions.assertEquals(newBank.getCreditCommission(), creditCommission);
        Assertions.assertEquals(newBank.getDebitCommission(), debitCommission);
        Assertions.assertEquals(newBank.getDoubtfulClientLimit(), doubtfulClientLimit);
    }

    /**
     * Testing if creating client by ClientBuilder works correctly(correct name, surname, passport info, address) and if client was added to bank system.
     */
    @Test
    public void creatingClientAndAddingToBank() {
        final String bankName = "Bank2";
        final double depositInterest = 10;
        final double creditCommission = 100;
        final double debitCommission = 5;
        final double doubtfulClientLimit = 50000;

        Bank bank = centralBank.createNewBank(bankName, depositInterest, creditCommission, debitCommission, doubtfulClientLimit);

        final String client_name = "Ivan";
        final String client_surname = "Ivanov";

        ClientBuilder clientBuilder = new ClientBuilder();
        clientBuilder.setFullName(new FullName(client_name, client_surname));
        Client client = clientBuilder.createClient();

        bank.addNewClient(client);

        Assertions.assertEquals(client_name, client.getFullName().getName());
        Assertions.assertEquals(client_surname, client.getFullName().getSurname());
        Assertions.assertNull(client.getPassportData());
        Assertions.assertNull(client.getAddress());
        Assertions.assertTrue(bank.getClients().contains(client));
    }

    /**
     * Testing if creating different account works correctly and client has all created account(debit, deposit, credit).
     */
    @Test
    public void creatingClientsAccounts() {
        final String bank_name = "Bank3";
        final double depositInterest = 10;
        final double creditCommission = 100;
        final double debitCommission = 5;
        final double doubtfulClientLimit = 50000;

        Bank bank = centralBank.createNewBank(bank_name, depositInterest, creditCommission, debitCommission, doubtfulClientLimit);

        final String client_name = "Ivan";
        final String client_surname = "Ivanov";

        ClientBuilder clientBuilder = new ClientBuilder();
        clientBuilder.setFullName(new FullName(client_name, client_surname));
        Client client = clientBuilder.createClient();

        bank.addNewClient(client);

        var deposit_period = Duration.ofDays(60);
        final double money_amount = 100000;
        DebitAccount debitAccount = bank.createDebitAccount(client.getId());
        CreditAccount creditAccount = bank.createCreditAccount(client.getId());
        DepositAccount depositAccount = bank.createDepositAccount(client.getId(), deposit_period, money_amount);

        Assertions.assertTrue(client.getAccounts().contains(debitAccount));
        Assertions.assertTrue(client.getAccounts().contains(creditAccount));
        Assertions.assertTrue(client.getAccounts().contains(depositAccount));
    }

    /**
     * Testing to replenish money to account, if we replenish money to account they actually appear there.
     */
    @Test
    public void replenishToAccount() {
        final String bankName = "Bank4";
        final double depositInterest = 10;
        final double creditCommission = 100;
        final double debitCommission = 5;
        final double doubtfulClientLimit = 50000;

        Bank bank = centralBank.createNewBank(bankName, depositInterest, creditCommission, debitCommission, doubtfulClientLimit);

        final String client_name = "Ivan";
        final String client_surname = "Ivanov";

        ClientBuilder clientBuilder = new ClientBuilder();
        clientBuilder.setFullName(new FullName(client_name, client_surname));
        Client client = clientBuilder.createClient();

        bank.addNewClient(client);

        final double money_amount = 100000;
        final double money_amount_twice = 200000;
        final double ZeroMoney = 0;
        DebitAccount debitAccount = bank.createDebitAccount(client.getId());

        Assertions.assertEquals(debitAccount.getMoney(), ZeroMoney);

        centralBank.replenishAccount(debitAccount.getId(), money_amount);
        Assertions.assertEquals(debitAccount.getMoney(), money_amount);

        centralBank.replenishAccount(debitAccount.getId(), money_amount);
        Assertions.assertEquals(debitAccount.getMoney(), money_amount_twice);
    }

    /**
     * Testing to withdraw from account, if we withdraw from account money should disappear from account.
     */
    @Test
    public void withdrawFromAccount() {
        final String bank_name = "Bank5";
        final double depositInterest = 10;
        final double creditCommission = 100;
        final double debitCommission = 5;
        final double doubtfulClientLimit = 50000;

        Bank bank = centralBank.createNewBank(bank_name, depositInterest, creditCommission, debitCommission, doubtfulClientLimit);

        final String client_name = "Ivan";
        final String client_surname = "Ivanov";

        ClientBuilder clientBuilder = new ClientBuilder();
        clientBuilder.setFullName(new FullName(client_name, client_surname));
        Client client = clientBuilder.createClient();

        bank.addNewClient(client);

        final double repl_money_amount = 100000;
        final double withdraw_money_amount = 40000;
        final double expected_money_left = repl_money_amount - withdraw_money_amount;

        DebitAccount debitAccount = bank.createDebitAccount(client.getId());
        centralBank.replenishAccount(debitAccount.getId(), repl_money_amount);

        centralBank.withdrawFromAccount(debitAccount.getId(), withdraw_money_amount);
        Assertions.assertEquals(expected_money_left, debitAccount.getMoney());
    }

    /**
     * Testing to transfer money from one account to another, from first account they should disappear and should appear at the second account.
     */
    @Test
    public void transferBetweenAccounts() {
        final String bank_name = "Bank6";
        final double depositInterest = 10;
        final double creditCommission = 100;
        final double debitCommission = 5;
        final double doubtfulClientLimit = 50000;

        Bank bank = centralBank.createNewBank(bank_name, depositInterest, creditCommission, debitCommission, doubtfulClientLimit);

        final String client1_name = "Ivan";
        final String client1_surname = "Ivanov";

        ClientBuilder client1Builder = new ClientBuilder();
        client1Builder.setFullName(new FullName(client1_name, client1_surname));
        Client client1 = client1Builder.createClient();

        final String client2_name = "Name";
        final String client2_surname = "Surname";

        ClientBuilder client2Builder = new ClientBuilder();
        client2Builder.setFullName(new FullName(client2_name, client2_surname));
        Client client2 = client2Builder.createClient();

        bank.addNewClient(client1);
        bank.addNewClient(client2);

        final double repl_money_amount = 100000;
        final double transfer_money_amount = 30000;
        final double expected_money_left = repl_money_amount - transfer_money_amount;

        DebitAccount debitAccount1 = bank.createDebitAccount(client1.getId());
        DebitAccount debitAccount2 = bank.createDebitAccount(client2.getId());
        centralBank.replenishAccount(debitAccount1.getId(), repl_money_amount);

        centralBank.transferMoneyBetweenAccounts(debitAccount1.getId(), debitAccount2.getId(), transfer_money_amount);
        Assertions.assertEquals(expected_money_left, debitAccount1.getMoney());
        Assertions.assertEquals(transfer_money_amount, debitAccount2.getMoney());
    }
}
